add turtle handler for paged resources

jwestgard opened the pull request:
This handler:
- works with a master graph of turtle metadata
- writes item-level turtle files to a metadata directory specified in the batch config
- trusts the hasPart predicate on each item to get the associated files
- creates sequence numbering beginning from 1 regardless of filename and enumerates the parts from there
- attaches item-level files (e.g. PDF of the entire letter) directly to the item resource
- writes a batch_index.yml describing the structure of the batch -- this index, if it already exists at runtime, is read as the basis for the batch

jwestgard committed:
add turtle handler for paged resources

peichman-umd commented on handler/turtle.py:
```diff
@@ -0,0 +1,274 @@
+                    try:
+                        id = os.path.basename(f).rstrip('.ttl')
+                        self.items[id] = self.create_item(fullpath)
+                    except:
```
Having this bare `except` here is simply swallowing any errors without logging any sort of explanation. It would be better to explicitly catch the anticipated exceptions (e.g., `KeyError` when `create_item` cannot find a file in the index) and issue a warning via the logger.

peichman-umd commented on handler/turtle.py:
```diff
@@ -0,0 +1,274 @@
+                else:
+                    parts[page_no]['files'].append(self.all_files[filename])
+            else:
+                print("ERROR!")
```
This should be a call to `logger.error` with a more expressive message.

peichman-umd commented on handler/turtle.py:
```diff
@@ -0,0 +1,274 @@
+        self.logger.info("Found {0} files".format(len(self.all_files)))
+
+        # Generate item-level metadata graphs and store as files
+        with open(self.metadata_file, 'r') as f:
```
This whole operation of parsing the metadata graph and creating the individual turtle file seems like the kind of thing that should be able to be controlled separately from the loading of each individual item, so that it only has to be run once on the batch, and not for every time the loader is run.

peichman-umd committed:
LIBFCREPO-495. Refactoring and streamlining of turtle handler.

* use logger and exceptions
* don't write intermediate metadata files; copy triples from master graph into a per-item source graph and read from there
* no longer generates the full batch index; just creates a file index once that maps basename of the file (used in dcterms:hasPart) to its full path
* remove dcterms:hasPart triples after reading data from the source graph

https://issues.umd.edu/browse/LIBFCREPO-495

peichman-umd commented:
@jwestgard These changes remove the need to configure a `BATCH_INDEX` or `METADATA_PATH` in the batch config YAML file. No intermediate TTL files are generated anymore; the `read_data()` method reads directly from the source graph to populate the item's data.

jwestgard commented:
DataReadException raised when non-unique filenames encountered is fatal.  Thus, multiple .DS_Store files will cause the batch to fail before loading anything.

jwestgard commented:
I have now done further testing and encountered duplicate file names in the following variations:

1. .DS_Store files
2. Thumbs.db files
3. duplicate filenames resulting from rework but not intended for inclusion in an object (e.g. new MD5 files found in a folder marked "corrected")
4. duplicate filenames caused by typos (/Volumes/UMD107/kap/data//phase1/KAP 7/hOCR/Box 07_Folder 01/litmss-023314-0001.xml and -0002.xml are errors for litmss-022314).

One approach to this problem might be to make the master file index include a list of paths for each unique filename, and to raise an exception when an object is created with a part that matches multiple paths.

jwestgard commented:
Noticing that files that are not TIF/JPG of the pages themselves are not included in the metadata turtle -- hence no hOCR or TXT transcriptions are being uploaded with the items.  Also, item-level files are not included (e.g. PDF of the complete letter).

peichman-umd commented:
The first two cases of duplicate filenames (.DS_Store and Thumbs.db) can easily be taken care of either by reinstating the filename regex or implementing an ignore list (much like .gitignore). For the rework and typos case, it seems less clear what to do, since the whole point is to allow a unique filename to be mapped to a path, regardless of how that path is structured.

peichman-umd commented:
Numbering the parts from 1 will be easy to put back in, it was an oversight when I was refactoring.

Is the hOCR files not in the turtle an upstream metadata issue for Bria to resolve?

peichman-umd committed:
LIBFCREPO-495. Renumber parts starting at 1.

https://issues.umd.edu/browse/LIBFCREPO-495

jwestgard commented:
To me it seems clear that non-unique files that are part of an object should be an exception, but otherwise the loader shouldn't care about them.  Hence, cases 1-3 shouldn't matter, but #4 should cause an exception (though not necessarily be fatal to the entire batch, since it is really only a problem when an item attempts to use it).

peichman-umd commented:
I'm not sure I understand case (3). Do you mean reworked TIF/JPG/etc files that have the same names as the original files, and that are included using `dcterms:hasPart`? Or are there files that are not getting included on the object that happen to have duplicate names (like the .DS_Store and Thumbs.db garbage)?

jwestgard commented:
This particular case is of a directory of files not included in any object, specifically a bunch of MD5 checksums.  It is possible that they are the result of rework where the file themselves were substituted with new files, but both sets of checksums were preserved for whatever reason.  They don't cause any problems because they aren't included in any objects, but because there are duplicated filenames, the loader balks.

jwestgard commented:
I thought if the file index recorded all paths to a particular filename, like this:
`litmss-999999-0001.jpg: ['path/to/first/litmss-999999-0001.jpg', 'path/to/second/litmss-999999-0001.jpg']`
Then at item creation time, one could simply check to make sure there was one and only one path. Zero or >1 paths results in a data read exception.  But that way we don't have to handle all possible cases of files that are irrelevant to objects being created.

peichman-umd committed:
LIBFCREPO-495. Record all files found in the file index.

Delay the check for unique filenames until the processing of an item.

https://issues.umd.edu/browse/LIBFCREPO-495

peichman-umd commented:
@jwestgard Point taken 😃

jwestgard commented:
We now have updated TTL with all files included thanks to @brialparker, so will do one last test and merge the PR and we should be good to go.

jwestgard commented:
Everything working as expected now.  Pages are numbered per letter, unused duplicate files being ignored, file index is written and reused when available, mapfile populated with unique ttl subject as identifier allowing reliable resumability.
